File: ios/joinery-kit/Sources/JoineryMemberKit/ConversationThreadView.swift
import SwiftUI
import JoineryKit

/// One conversation: message bubbles in a scroll with a compose bar pinned
/// to the bottom, cursor pagination on scroll, and a mute/delete menu.
/// Opening the thread marks it read server-side as a side effect of the
/// load call. New-conversation compose (`origin: .compose`) dedups to an
/// existing 1:1 conversation on load, or creates one on first send.
public struct ConversationThreadView: View {
    @StateObject private var store: ConversationThreadStore
    @Environment(\.dismiss) private var dismiss
    @State private var showDeleteConfirm = false
    @State private var deleteError: String?

    public init(client: APIClient, origin: ThreadOrigin) {
        _store = StateObject(wrappedValue: ConversationThreadStore(api: ConversationAPI(client: client), origin: origin))
    }

    public var body: some View {
        content
            .navigationTitle(store.otherDisplayName)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar { toolbarContent }
            .safeAreaInset(edge: .bottom) { composer }
            .task {
                if case .loading = store.phase { await store.load() }
            }
            .confirmationDialog("Delete this conversation?", isPresented: $showDeleteConfirm, titleVisibility: .visible) {
                Button("Delete", role: .destructive) {
                    Task {
                        do {
                            try await store.delete()
                            dismiss()
                        } catch {
                            deleteError = (error as? JoineryAPIError)?.displayMessage ?? "Could not delete the conversation."
                        }
                    }
                }
                Button("Cancel", role: .cancel) {}
            }
            .alert("Could not delete", isPresented: deleteErrorBinding) {
                Button("OK") {}
            } message: {
                Text(deleteError ?? "")
            }
    }

    @ViewBuilder
    private var content: some View {
        switch store.phase {
        case .loading:
            ProgressView()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .accessibilityIdentifier("conversation_thread_loading")
        case .failed(let message):
            VStack(spacing: 12) {
                Text(message)
                    .multilineTextAlignment(.center)
                    .foregroundStyle(.secondary)
                    .accessibilityIdentifier("conversation_thread_error")
                Button("Try Again") { Task { await store.load() } }
                    .buttonStyle(.borderedProminent)
            }
            .padding()
        case .loaded:
            transcript
        }
    }

    private var transcript: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 10) {
                    if store.hasMore {
                        HStack {
                            Spacer()
                            if store.isLoadingMore {
                                ProgressView()
                            } else {
                                Button("Load more") { Task { await store.loadMore() } }
                                    .font(.caption)
                            }
                            Spacer()
                        }
                    }
                    if store.messages.isEmpty {
                        emptyState
                    }
                    ForEach(store.messages) { message in
                        MessageBubble(message: message)
                            .id(message.id)
                    }
                    Color.clear.frame(height: 1).id(bottomAnchor)
                }
                .padding(.horizontal, 14)
                .padding(.top, 12)
            }
            .accessibilityIdentifier("conversation_transcript")
            .onChange(of: store.messages.count) { _ in
                withAnimation(.easeOut(duration: 0.2)) { proxy.scrollTo(bottomAnchor, anchor: .bottom) }
            }
            .onAppear { proxy.scrollTo(bottomAnchor, anchor: .bottom) }
        }
    }

    private var emptyState: some View {
        Text("Say hello.")
            .foregroundStyle(.secondary)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 60)
            .accessibilityIdentifier("conversation_thread_empty")
    }

    private var composer: some View {
        HStack(alignment: .bottom, spacing: 8) {
            TextField("Message", text: $store.composerText, axis: .vertical)
                .textFieldStyle(.plain)
                .lineLimit(1...5)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(RoundedRectangle(cornerRadius: 20).fill(Color(.secondarySystemBackground)))
                .accessibilityIdentifier("conversation_composer")
            Button {
                Task { await store.send() }
            } label: {
                if store.isSending {
                    ProgressView().frame(width: 32, height: 32)
                } else {
                    Image(systemName: "arrow.up.circle.fill")
                        .font(.system(size: 32))
                }
            }
            .disabled(!store.canSend)
            .accessibilityIdentifier("conversation_send")
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(.bar)
    }

    @ToolbarContentBuilder
    private var toolbarContent: some ToolbarContent {
        ToolbarItem(placement: .topBarTrailing) {
            Menu {
                if store.conversationID != nil {
                    Button {
                        Task { await store.setMuted(!store.isMuted) }
                    } label: {
                        Label(store.isMuted ? "Unmute" : "Mute", systemImage: store.isMuted ? "bell" : "bell.slash")
                    }
                    Button(role: .destructive) {
                        showDeleteConfirm = true
                    } label: {
                        Label("Delete", systemImage: "trash")
                    }
                }
            } label: {
                Image(systemName: "ellipsis.circle")
            }
            .accessibilityIdentifier("conversation_thread_menu")
            .disabled(store.conversationID == nil)
        }
    }

    private let bottomAnchor = "conversation_bottom_anchor"

    private var deleteErrorBinding: Binding<Bool> {
        Binding(get: { deleteError != nil }, set: { if !$0 { deleteError = nil } })
    }
}

/// One message bubble: the caller's on the right, the other participant's
/// on the left.
struct MessageBubble: View {
    let message: ThreadMessage

    var body: some View {
        HStack {
            if message.isMine { Spacer(minLength: 40) }
            VStack(alignment: message.isMine ? .trailing : .leading, spacing: 3) {
                Text(message.body)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 9)
                    .background(
                        RoundedRectangle(cornerRadius: 18)
                            .fill(message.isMine ? Color.accentColor : Color(.secondarySystemBackground))
                    )
                    .foregroundStyle(message.isMine ? .white : .primary)
                    .textSelection(.enabled)
                Text(MemberDisplay.dateTimeLabel(message.time))
                    .font(.caption2)
                    .foregroundStyle(.tertiary)
            }
            .accessibilityIdentifier(message.isMine ? "conversation_message_mine" : "conversation_message_theirs")
            if !message.isMine { Spacer(minLength: 40) }
        }
    }
}

File: android/joinery-android-member/src/main/java/com/getjoinery/memberkit/ConversationThreadView.kt
package com.getjoinery.memberkit

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material.icons.outlined.Notifications
import androidx.compose.material.icons.outlined.NotificationsOff
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.getjoinery.android.JoineryApiError
import kotlinx.coroutines.launch

/**
 * One conversation: message bubbles in a scroll with a compose bar pinned to
 * the bottom, cursor pagination on scroll, and a mute/delete menu. Opening the
 * thread marks it read server-side as a side effect of the load call.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ConversationThreadView(
    api: ConversationApi,
    origin: ThreadOrigin,
    onClose: () -> Unit,
) {
    val scope = rememberCoroutineScope()
    val store = remember { ConversationThreadStore(api, origin) }
    var showDeleteConfirm by remember { mutableStateOf(false) }
    var deleteError by remember { mutableStateOf<String?>(null) }
    var menuOpen by remember { mutableStateOf(false) }

    BackHandler(onBack = onClose)
    LaunchedEffect(Unit) {
        if (store.phase is MemberPhase.Loading) store.load()
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text(store.otherDisplayName, maxLines = 1, overflow = TextOverflow.Ellipsis) },
                navigationIcon = {
                    IconButton(onClick = onClose) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
                actions = {
                    if (store.conversationId != null) {
                        IconButton(onClick = { menuOpen = true }, modifier = Modifier.testTag("member_conversation_thread_menu")) {
                            Icon(Icons.Filled.MoreVert, contentDescription = "More actions")
                        }
                        DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                            DropdownMenuItem(
                                text = { Text(if (store.isMuted) "Unmute" else "Mute") },
                                leadingIcon = {
                                    Icon(
                                        if (store.isMuted) Icons.Outlined.Notifications else Icons.Outlined.NotificationsOff,
                                        contentDescription = null,
                                    )
                                },
                                onClick = {
                                    menuOpen = false
                                    scope.launch { store.setMuted(!store.isMuted) }
                                },
                            )
                            DropdownMenuItem(
                                text = { Text("Delete") },
                                leadingIcon = {
                                    Icon(Icons.Outlined.Delete, contentDescription = null, tint = MaterialTheme.colorScheme.error)
                                },
                                onClick = { menuOpen = false; showDeleteConfirm = true },
                            )
                        }
                    }
                },
            )
        },
        bottomBar = { Composer(store) },
    ) { padding ->
        when (val phase = store.phase) {
            is MemberPhase.Loading ->
                Box(Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(Modifier.testTag("member_conversation_thread_loading"))
                }
            is MemberPhase.Failed ->
                RetryBox(phase.message, "member_conversation_thread_error", "member_conversation_thread_retry", Modifier.padding(padding)) {
                    scope.launch { store.load() }
                }
            is MemberPhase.Loaded ->
                Transcript(store, Modifier.padding(padding))
        }
    }

    if (showDeleteConfirm) {
        AlertDialog(
            onDismissRequest = { showDeleteConfirm = false },
            title = { Text("Delete this conversation?") },
            confirmButton = {
                TextButton(onClick = {
                    showDeleteConfirm = false
                    scope.launch {
                        try {
                            store.delete()
                            onClose()
                        } catch (e: Exception) {
                            deleteError = (e as? JoineryApiError)?.displayMessage ?: "Could not delete the conversation."
                        }
                    }
                }) { Text("Delete") }
            },
            dismissButton = { TextButton(onClick = { showDeleteConfirm = false }) { Text("Cancel") } },
        )
    }

    deleteError?.let { message ->
        AlertDialog(
            onDismissRequest = { deleteError = null },
            title = { Text("Could not delete") },
            text = { Text(message) },
            confirmButton = { TextButton(onClick = { deleteError = null }) { Text("OK") } },
        )
    }

    store.sendError?.let { message ->
        AlertDialog(
            onDismissRequest = { store.clearSendError() },
            title = { Text("Could not send") },
            text = { Text(message) },
            confirmButton = { TextButton(onClick = { store.clearSendError() }) { Text("OK") } },
        )
    }
}

@Composable
private fun Transcript(store: ConversationThreadStore, modifier: Modifier = Modifier) {
    val scope = rememberCoroutineScope()
    val listState = rememberLazyListState()

    // Keep the newest message in view as the transcript grows.
    LaunchedEffect(store.messages.size) {
        if (store.messages.isNotEmpty()) listState.animateScrollToItem(store.messages.size - 1)
    }

    LazyColumn(
        state = listState,
        modifier = modifier.fillMaxSize().testTag("member_conversation_transcript"),
        verticalArrangement = Arrangement.spacedBy(10.dp),
        contentPadding = androidx.compose.foundation.layout.PaddingValues(horizontal = 14.dp, vertical = 12.dp),
    ) {
        if (store.hasMore) {
            item {
                Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                    if (store.isLoadingMore) {
                        CircularProgressIndicator(Modifier.size(20.dp))
                    } else {
                        TextButton(onClick = { scope.launch { store.loadMore() } }) { Text("Load more") }
                    }
                }
            }
        }
        if (store.messages.isEmpty()) {
            item {
                Text(
                    "Say hello.",
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth().padding(vertical = 60.dp).testTag("member_conversation_thread_empty"),
                )
            }
        }
        items(store.messages.size, key = { store.messages[it].messageId }) { index ->
            MessageBubble(store.messages[index])
        }
    }
}

@Composable
private fun MessageBubble(message: ThreadMessage) {
    Row(Modifier.fillMaxWidth()) {
        if (message.isMine) Spacer(Modifier.weight(1f).sizeIn(minWidth = 40.dp))
        Column(
            horizontalAlignment = if (message.isMine) Alignment.End else Alignment.Start,
            verticalArrangement = Arrangement.spacedBy(3.dp),
            modifier = Modifier.testTag(if (message.isMine) "member_conversation_message_mine" else "member_conversation_message_theirs"),
        ) {
            Surface(
                shape = RoundedCornerShape(18.dp),
                color = if (message.isMine) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.surfaceVariant,
            ) {
                SelectionContainer {
                    Text(
                        message.body,
                        color = if (message.isMine) MaterialTheme.colorScheme.onPrimary else MaterialTheme.colorScheme.onSurface,
                        modifier = Modifier.padding(horizontal = 14.dp, vertical = 9.dp),
                    )
                }
            }
            Text(
                MemberDisplay.dateTimeLabel(message.time),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        if (!message.isMine) Spacer(Modifier.weight(1f).sizeIn(minWidth = 40.dp))
    }
}

@Composable
private fun Composer(store: ConversationThreadStore) {
    val scope = rememberCoroutineScope()
    Surface(tonalElevation = 3.dp) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.Bottom,
        ) {
            OutlinedTextField(
                value = store.composerText,
                onValueChange = { store.composerText = it },
                placeholder = { Text("Message") },
                maxLines = 5,
                shape = RoundedCornerShape(20.dp),
                modifier = Modifier.weight(1f).testTag("member_conversation_composer"),
            )
            IconButton(
                onClick = { scope.launch { store.send() } },
                enabled = store.canSend,
                modifier = Modifier.testTag("member_conversation_send"),
            ) {
                if (store.isSending) {
                    CircularProgressIndicator(Modifier.size(24.dp), strokeWidth = 2.dp)
                } else {
                    Icon(
                        Icons.AutoMirrored.Filled.Send,
                        contentDescription = "Send",
                        tint = if (store.canSend) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }
    }
}
